From c8a4d58bdf732f96009033233bd8e96841df6b99 Mon Sep 17 00:00:00 2001
From: Jianwei Fan <jianwei.fan@rock-chips.com>
Date: Wed, 17 Jan 2024 06:21:40 +0000
Subject: [PATCH 02/11] media: i2c: rk628: set default timings when query
 timing if hdmi unplug

Type: Fix
Redmine ID: #460955
Associated modifications: N/A
Test: pass

Change-Id: Ia88eb29b8f628ad141a256e9cb9229d9b6491851
Signed-off-by: Jianwei Fan <jianwei.fan@rock-chips.com>
Signed-off-by: Guochun Huang <hero.huang@rock-chips.com>
---
 drivers/media/i2c/rk628/rk628_bt1120_v4l2.c |  8 ++++++++
 drivers/media/i2c/rk628/rk628_csi_v4l2.c    | 11 ++++++++++-
 2 files changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c b/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
index f02022e4ac65bb..9cb0b8ba74616c 100644
--- a/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
+++ b/drivers/media/i2c/rk628/rk628_bt1120_v4l2.c
@@ -1216,7 +1216,15 @@ static int rk628_bt1120_query_dv_timings(struct v4l2_subdev *sd,
 		struct v4l2_dv_timings *timings)
 {
 	int ret;
+	struct rk628_bt1120 *bt1120 = to_bt1120(sd);
+	struct v4l2_dv_timings default_timing =
+				V4L2_DV_BT_CEA_640X480P59_94;
 
+	if (!tx_5v_power_present(sd) || bt1120->nosignal) {
+		*timings = default_timing;
+		v4l2_info(sd, "%s: not detect 5v, set default timing\n", __func__);
+		return 0;
+	}
 	ret = rk628_bt1120_get_detected_timings(sd, timings);
 	if (ret)
 		return ret;
diff --git a/drivers/media/i2c/rk628/rk628_csi_v4l2.c b/drivers/media/i2c/rk628/rk628_csi_v4l2.c
index bc58838df33ea3..d6c0ebf6d7592d 100644
--- a/drivers/media/i2c/rk628/rk628_csi_v4l2.c
+++ b/drivers/media/i2c/rk628/rk628_csi_v4l2.c
@@ -1786,7 +1786,14 @@ static int rk628_csi_query_dv_timings(struct v4l2_subdev *sd,
 {
 	int ret;
 	struct rk628_csi *csi = to_csi(sd);
+	struct v4l2_dv_timings default_timing =
+				V4L2_DV_BT_CEA_640X480P59_94;
 
+	if (!tx_5v_power_present(sd) || csi->nosignal) {
+		*timings = default_timing;
+		v4l2_info(sd, "%s: not detect 5v, set default timing\n", __func__);
+		return 0;
+	}
 	mutex_lock(&csi->confctl_mutex);
 	ret = rk628_csi_get_detected_timings(sd, timings);
 	mutex_unlock(&csi->confctl_mutex);
@@ -2435,10 +2442,12 @@ static irqreturn_t plugin_detect_irq(int irq, void *dev_id)
 {
 	struct rk628_csi *csi = dev_id;
 	struct v4l2_subdev *sd = &csi->sd;
+	bool plugin;
 
 	/* control hpd after 50ms */
 	schedule_delayed_work(&csi->delayed_work_enable_hotplug, HZ / 20);
-	tx_5v_power_present(sd);
+	plugin = tx_5v_power_present(sd);
+	v4l2_info(sd, "%s: 5v_det:%d\n", __func__, plugin);
 
 	return IRQ_HANDLED;
 }
-- 
2.17.1

